README.md
Mensa-Dresden
=============

Perl interface to receive the offering of a mensa in Dresden

Changes
Revision history for Perl extension Mensa::Dresden.

0.01  Sun Sep 23 19:11:56 2012
	- original version; created by h2xs 1.23



lib/Mensa/Dresden/Meal.pm
package Mensa::Dresden::Meal;

use 5.014002;
use strict;
use warnings;

=head1 NAME

Mensa::Dresden::Meal - The datastructure which contains all
meal-related information

=head1 SYNOPSIS

  use Mensa::Dresden::Meal;

  $meal_xml = XML::LibXML->load_xml(string => <<XML);
  <?xml version="1.0" encoding="UTF-8" ?>
  <offering>
   	  <mensa name="$mensa_name">
          <meal url="$url">
              <name>$name</name>
              <ingredient>$ingredient_1</ingredient>
              <ingredient>$ingredient_2</ingredient>
          </meal>
      </mensa>
  </offering>
  XML

  $meal = Mensa::Dresden::Meal->new($meal_xml);

=head1 DESCRIPTION

This datastructure contains all meal relevant information. It is
used to deliver the offering in a well defined structure which
is applicable to be filtered.

=head2 EXPORT

None by default. The C<all>-tag imports the ingredient-constants
(see section CONSTANTS).

=cut

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = ( 'all' => [ qw(
	PORK BEEF VEGETARIAN ALCOHOL GARLIC VEGAN
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our $VERSION = '0.01';


use Carp;
use Fcntl 'SEEK_SET';
use XML::LibXML;
use XML::LibXSLT;

=head2 CONSTANTS

Following constants can be used to filter the meals.

=over 4

=item B<PORK>

Constant indicating the meal contains pork.

=item B<BEEF>

Constant indicating the meal contains beef.

=item B<VEGETARIAN>

Constant indicating the meal is vegetarian.

=item B<ALCOHOL>

Constant indicating the meal contains alcohol.

=item B<GARLIC>

Constant indicating the meal contains garlic.

=item B<VEGAN>

Constant indicating the meal is vegan.

=back

=cut

use constant {
	PORK => 'Schweinefleisch',
	BEEF => 'Rindfleisch',
	VEGETARIAN => 'kein Fleisch',
	ALCOHOL => 'Alkohol',
	GARLIC => 'Knoblauch',
	VEGAN => 'vegan'
};


my $schema = XML::LibXML::Schema->new(string => join ('', <DATA>));

use subs qw(validate get_name get_ingredients);

=head2 METHODS

=over 4

=item B<new>

Instantiates a new meal by extracting all required information
from the given XML node.

=cut

sub new {
	my $class = shift;
	my $xml = shift;
#	validate($xml);
	my $self = {
		url => $xml->getAttribute('url'),
		name => get_name($xml),
		ingredients => [ get_ingredients($xml) ]
	};
	return bless $self, $class;
}

#
# Validates the given XML against the schema.
#
sub validate($) {
	my $xml = shift;
	eval { $schema->validate($xml) };
	croak("No valid meal XML: $@") if $@;
}

#
# Extracts the text content from the given XML node.
#
sub get_name($) {
	my $xml = shift;
	my ($element) = $xml->getChildrenByTagName('name');
	return $element->textContent();
}

#
# Extracts all ingredients from the given XML node.
#
sub get_ingredients($) {
	my $xml = shift;
	return map {
		$_->textContent()
	} $xml->getElementsByTagName('ingredient');
}

=item B<name>

Returns the name of the meal.

=cut

sub name {
	my $self = shift;
	return $self->{name};
}

=item B<url>

Returns the URL to the meal's detail-site.

=cut

sub url {
	my $self = shift;
	return $self->{url};
}

=item B<ingredients>

Returns the ingredients of the meal as list.

=cut

sub ingredients() {
	my $self = shift;
	return @{ $self->{ingredients} };
}

=item B<to_string>

Returns a printable string which contains the information of the meal
as follows:

  name of the meal
  > ingredients
  http://example.org/meal's-detail-site

=cut

sub to_string() {
	my $self = shift;
	my $string = '';
	$string .= $self->name . "\n";
	my @ingredients = $self->ingredients;
	@ingredients = '-' unless @ingredients;
	local $" = ', ';
	$string .= "> @ingredients\n";
	$string .= $self->url;
	return $string;
}

=back

=cut


1;

=head1 AUTHOR

8ware, E<lt>8wared@googlemail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by 8ware

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.2 or,
at your option, any later version of Perl 5 you may have available.

=cut

__DATA__
<?xml version="1.0" encoding="UTF-8" ?>
<xs:schema
		xmlns:xs="http://www.w3.org/2001/XMLSchema"
		elementFormDefault="qualified"
		version="0.1">

	<xs:complexType name="Offering">
		<xs:sequence>
			<xs:element ref="mensa" maxOccurs="unbounded"/>
		</xs:sequence>
	</xs:complexType>

	<xs:complexType name="Mensa">
		<xs:sequence>
			<xs:element name="meal" type="Meal" minOccurs="0" maxOccurs="unbounded"/>
		</xs:sequence>
		<xs:attribute name="name" type="xs:string" use="required"/>
	</xs:complexType>

	<xs:complexType name="Meal">
		<xs:sequence>
			<xs:element name="name" type="xs:string"/>
		</xs:sequence>
		<xs:attribute name="url" type="xs:anyURI"/>
	</xs:complexType>

	<xs:element name="offering" type="Offering"/>
	<xs:element name="mensa" type="Mensa"/>
	<xs:element name="meal" type="Meal"/>

</xs:schema>



lib/Mensa/Dresden/Filter.pm
package Mensa::Dresden::Filter;

use 5.014002;
use strict;
use warnings;

=head1 NAME

Mensa::Dresden::Filter - Utility-class to filter out meals

=head1 SYNOPSIS

  use Mensa::Dresden::Filter ':all';

  $steak_filter = Mensa::Dresden::Filter->new(
      name => qr/steak/i    # or
  #   NAME, qr/steak/i      # no typos
  );
  
  $anti_vegan_filter = Mensa::Dresden::Filter->new(
      ingredients => qr/vegan/, NEGATIVE    # or
  #   INGREDIENTS, qr/vegan/, NEGATIVE      # no typos
  );

=head1 DESCRIPTION

This utility-class serves as filter for meals. A filter consists
of a criterion, a regular expression and a flag which impacts the
filter's behavior. As the filter is applied via the C<pass>-method
the criterion is evaluated against the expression, while the flag
effects the boolean result.

=head2 EXPORT

None by default. The C<all>-tag exports the constants which are
listed in the CONSTANTS section. 

=cut

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = ( all => [ qw(
	NAME INGREDIENTS
	POSITIVE NEGATIVE
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our $VERSION = '0.01';


use Carp;

=head2 CONSTANTS

=over 4

=item B<NAME>

Represents the C<name>-criterion which applies the filter to the
name of the meal.

=item B<INGREDIENTS>

Represents the C<ingredients>-criterion which applies the filter
to each ingredient of the meal.

=item B<POSITIVE>

Indicates that the filter let only meals pass if the criterion
matches the expression. Because this behavior is the default one
this flag does not impact the filter's functionality.

=item B<NEGATIVE>

Indicates that the filter is inverted and only meals can pass
which criterion does not match the expression.

=back

=cut

use constant {
	NAME => 'name',
	INGREDIENTS => 'ingredients',

	POSITIVE => 0,
	NEGATIVE => 1
};

=head2 METHODS

=over 4

=item B<new>

Accepts the filter-criterion (e.g. C<name> or C<ingredients>), a regular
expression and optionally the inversion-flag (see constants) in this order.
The regular expression is wrapped with the regex-quote and the ignore-case
flag if not happened. An already interpolated expression (using the
C<qr>-operator) will exactly behave as passed to this method, due the regex
is not interpolated twice.

=cut

sub new {
	my $class = shift;
	my $criterion = shift;
	croak("Unsupported criterion: $criterion")
			unless $criterion eq NAME or $criterion eq INGREDIENTS;
	my $regex = shift;
	my $negative = shift || 0; # defined ?
	my $self = {
		criterion => $criterion,
		regex => qr/$regex/i,
		negative => $negative ? 1 : 0 # enforce 1 or 0
	};
	return bless $self, $class;
}

=item B<pass>

Accepts an instance of Mensa::Dresden::Meal and returns C<true> if this
meal has passed the filter.

=cut

sub pass($) {
	my $self = shift;
	my $meal = shift;
	my $criterion = $self->{criterion};
	my $regex = $self->{regex};
	my $negative = $self->{negative};
	my $result = 0;
	for ($meal->$criterion) {
		$result |= /$regex/;
	}
	return $negative ^ $result;
}

=item B<is_negative>

Returns C<true> if the negative-/inversion-flag was set;
C<false> otherwise.

=cut

sub is_negative() {
	my $self = shift;
	return $self->{negative};
}

=back

=cut


1;
__END__
=head1 AUTHOR

8ware, E<lt>8wared@googlemail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by 8ware

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.2 or,
at your option, any later version of Perl 5 you may have available.

=cut



lib/Mensa/Dresden.pm
package Mensa::Dresden;

use 5.014002;
use strict;
use warnings;

=head1 NAME

Mensa::Dresden - Perl interface to receive offerings of Dresden's canteens

=head1 SYNOPSIS

  use Mensa::Dresden ':all';

  $steak_filter = create_filter(name => qr/steak/i);
  $anti_tofu_filter = create_filter(name => qr/tofu/i, NEGATIVE);

  $mensa = Mensa::Dresden->new('Alte Mensa', $steak_filter);

  $antivegan = $mensa->create_filter(ingredients => qr/vegan/, NEGATIVE);
  $mensa->add_filter($anti_tofu_filter);

  @meals = $mensa->get_offering();
  @meals = $mensa->get_offering(TOMORROW);
  @meals = $mensa->get_offering(MONDAY, NEXT_WEEK);

=head1 DESCRIPTION

This module provides a simple interface to receive the offerings of
Dresden's canteens. Because tastes differ some filters can be specified
to eliminate loathsome meals. The mensa-script which comes along with
this distribution implements a simple command line interface to check
the canteens offering.

=head2 EXPORT

None by default, but there are several constants and some methods which can
be imported by the C<all>-tag. Additionally all date-related constants are
exported if the C<date>-tag is specified. Similar, all filter-related stuff
is provided if the C<filter>-tag is given. The next section (CONSTANTS) shows
which constants are available.

=cut

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = (
	all => [ qw(
		TODAY TOMORROW
		MONDAY TUESDAY WEDNESDAY THURSDAY FRIDAY SATURDAY SUNDAY
		THIS_WEEK NEXT_WEEK AFTERNEXT_WEEK
		create_filter NAME INGREDIENTS POSITIVE NEGATIVE
		PORK BEEF VEGETARIAN ALCOHOL GARLIC VEGAN
	) ],
	date => [ qw(
		TODAY TOMORROW
		MONDAY TUESDAY WEDNESDAY THURSDAY FRIDAY SATURDAY SUNDAY
		THIS_WEEK NEXT_WEEK AFTERNEXT_WEEK
	) ],
	filter => [ qw(
		create_filter NAME INGREDIENTS POSITIVE NEGATIVE
		PORK BEEF VEGETARIAN ALCOHOL GARLIC VEGAN
	) ]
);

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our $VERSION = '0.02';


use Carp;
use Fcntl 'SEEK_SET';
use HTTP::Request;
use LWP::UserAgent;
use XML::LibXML;
use XML::LibXSLT;

use Mensa::Dresden::Filter ':all';
use Mensa::Dresden::Meal ':all';

=head2 CONSTANTS

=head3 FILTER-RELATED

These constants are useful to create filters. See the description
of the C<create_filter>-method and the C<Dresden::Mensa::Filter>
module for further information.

=over 4

=item B<NAME>

The identifier for name-specific filters.

=item B<INGREDIENTS>

The identifier for ingredient-specific filters. Following constants
can be used to specify an ingredient-filter:

=over 8

=item B<PORK>

Constant indicating the meal contains pork.

=item B<BEEF>

Constant indicating the meal contains beef.

=item B<VEGETARIAN>

Constant indicating the meal is vegetarian.

=item B<ALCOHOL>

Constant indicating the meal contains alcohol.

=item B<GARLIC>

Constant indicating the meal contains garlic.

=item B<VEGAN>

Constant indicating the meal is vegan.

=back

=item B<POSITIVE>

This flag has no effect on the filter but is given for completeness.

=item B<NEGATIVE>

This flag indicates that the filter will be inverted.

=back

=head3 DATE-RELATED

The date related constants serve as a more readable form to specify
what meals are requested of which date, relative to the today's date.
See the desciption of the C<get_offering>-method for more information.

=over 4

=item B<TODAY>

Represents the today's day-number. As the C<get_offering>-method
supposes the current day if the arguments are omitted, this constant
is listed only for completeness.

=item B<TOMORROW>

Represents the tomorrow's day-number.

=item B<MONDAY> B<TUESDAY> B<WEDNESDAY> B<THURSDAY> B<FRIDAY>
      B<SATURDAY> B<SUNDAY>

Represents the numbers of the denoted days.

=item B<THIS_WEEK> B<NEXT_WEEK> B<AFTERNEXT_WEEK>

Represents the numbers for the current, next and after next week.

=back

=cut

use constant {
	TODAY => (localtime time)[6],
	TOMORROW => ((localtime time)[6] +1) % 7,

	MONDAY => 1,
	TUESDAY => 2,
	WEDNESDAY => 3,
	THURSDAY => 4,
	FRIDAY => 5,
	SATURDAY => 6,
	SUNDAY => 0,

	THIS_WEEK => 0,
	NEXT_WEEK => 1,
	AFTERNEXT_WEEK => 2
};


#
# The base URL of the canteen offerings.
#
our $URL = 'http://www.studentenwerk-dresden.de/mensen/speiseplan/';

#
# The canteen-names with their appropriate mensa-IDs.
#
our %MENSAS = (
	'Neue Mensa' => 8,
	'Alte Mensa' => 18,
	'Mensa Reichenbachstraße' => 9,
	'Mensologie' => 5,
	'Mensa Siedepunkt' => 6,
	'Mensa Johannstadt' => 32,
	'Mensa Blau' => 12,
	'BioMensa U-Boot' => undef, # no ID given, yet
	'Mensa TellerRandt' => 7,
	'Mensa Zittau' => 1,
	'Mensa Stimm-Gabel' => 13,
	'Mensa Palucca Schule' => 14,
	'Mensa Görlitz' => 15,
	'Mensa Haus VII' => 16,
	'Mensa Sport' => 19,
	'Mensa Kreuzgymnasium' => 20
);


my $stylesheet = XML::LibXSLT->new->parse_stylesheet(
	XML::LibXML->load_xml(IO => *DATA)
);

use subs qw(get_url fetch_html load_stylesheet filter filter_with);

=head2 METHODS

=over 4

=item B<new>

Creates a new instance of a Dresden's canteen. The first argument is the
canteen's name while all succeeding arguments must be filters. See the
C<create_filter>- and C<add_filter>-method for more information.

=cut

sub new {
	my $class = shift;
	my $mensa = shift;
	my @filters = @_;
	my @mensas = keys %MENSAS;
	return bless {
		name => $mensa,
		p_filters => [ grep { not $_->is_negative } @filters ],
		n_filters => [ grep { $_->is_negative } @filters ]
	}, $class if $mensa ~~ @mensas;
	croak("Unknown mensa: $mensa");
}

=item B<get_name>

Returns the name of the canteen.

=cut

sub get_name() {
	my $self = shift;
	return $self->{name};
}

#
# The constants which indicate the usage of positive and negative filters.
# Intended for a more readable usage of the 'filter'-method.
#
use constant {
	positive => 0,
	negative => 1
};

=item B<get_offering>

Accepts two arguments. The first one is the day, where 0 indicates sunday and
3 indicates wednesday. The second argument is the week, where 0 represents the
current, 1 the next and 2 the after next week. If both is omitted the current
day and week is supposed while the current week is assumed if only the week
parameter is omitted. After the offering was scrapped, the meals are filtered
as follows:

=cut

sub get_offering(;$$) {
	my $self = shift;
	my $day = defined $_[0] ? shift : (localtime time)[6];
	my $week = defined $_[0] ? shift : 0;
#	my $stylesheet = load_stylesheet();
	my $url = get_url($week, $day);
	my $html = fetch_html($url);
	my $offering = $stylesheet->transform($html,
		XML::LibXSLT::xpath_to_string(name => $self->{name}),
	);
	my @meals;
	for ($offering->getElementsByTagName('meal')) {
		my $meal = Mensa::Dresden::Meal->new($_);
		push @meals, $meal;
	}
	return $self->filter(@meals);
}

#
# Delivers the URL which is dependent on the given day and week.
#
sub get_url($$) {
	my ($week, $day) = @_;
	croak("Not a valid value: week=$week") if $week < 0 or $week > 2;
	croak("Not a valid value: day=$day") if $day < 0 or $day > 6;
	return $URL . "w$week-d$day.html";
}

#
# Fetches the HTML resource from the given URL and returns it as
# XML document.
#
sub fetch_html($) {
	my $url = shift;
	my $agent = LWP::UserAgent->new();
	my $request = HTTP::Request->new(GET => $url);
	my $response = $agent->request($request);
	croak("Received ".$response->status_line()) unless $response->is_success();
	my $html = XML::LibXML->load_html(
		string => $response->content(),
		load_ext_dtd => 0,
		expand_entities => 1,
		recover => 2,
		suppress_warnings => 1,
		suppress_errors => 1
	);
	return $html;
}

#
# Loads the stylesheet which is used to transform the HTML resource
# into a more appropriate XML representation.
#
sub load_stylesheet() {
	my $xslt = XML::LibXSLT->new();
	my $position = tell DATA;
	my $xsl = XML::LibXML->load_xml(IO => *DATA);
	seek DATA, $position, SEEK_SET;
	my $stylesheet = $xslt->parse_stylesheet($xsl);
	return $stylesheet;
}

#
# Filters the given meals. The filter mechanism works as follows:
# 1. Apply all positive filters to all meals, if at least one matches
#    add the meal to the list of filtered ones.
# 2. Apply all negative filters to the already filtered meals, if at
#    least one does not match, remove the meal, i.e. don't add it.
# 3. If no meals were left after step 1 and 2 apply all negative
#    filters to all meals and add the meal only of if all filtern match
# 4. If still no meals are left, return all meals.
#
sub filter(@) {
	my $self = shift;
	my @meals = @_;
	my @filtered_meals;
	@filtered_meals = $self->filter_with(positive, @meals);
	@filtered_meals = $self->filter_with(negative, @filtered_meals);
	return @filtered_meals if @filtered_meals;
	@filtered_meals = $self->filter_with(negative, @meals);
	return @filtered_meals if @filtered_meals;
	return @meals;
}

#
# Applies the filter of the mensa to the given meals. The first argument
# indicates whether to use the positive or negative filters. If true the
# negative filters are used. The behavior of positive and negative filters
# differ. If at least one positive filter matches, the meal passes and
# will be added to the filtered-array. Otherwise all negative filters must
# match to retain the meal.
#
sub filter_with($@) {
	my $self = shift;
	my $negative = shift;
	my @meals = @_;
	my @filters = @{ $self->{($negative ? 'n' : 'p').'_filters'} };
	return @meals unless @filters;
	my @filtered;
	for my $meal (@meals) {
		my $matches = grep { $_->pass($meal) } @filters;
		push @filtered, $meal
				if not $negative and $matches or $matches == @filters;
	}
	return @filtered;
}

=item B<create_filter>

Creates a new filter. The first argument is the criterion which should be
filtered, i.e. one of C<name> or C<ingredients>. To avoid typos the
C<NAME>- and C<INGREDIENTS>-constant should be used for identification.
The second argument is the regular expression which the criterion must match
to get not rejected. The third parameter is optional and indicates whether
the filter is positive or negative, i.e. a negative filter let the meal pass
only if the criterion does not match the expression. To keep it more readable
one can use the C<NEGATIVE>-constant to get this done (the C<POSITIVE>-constant
can also be used, but is implicitly supposed if omitted). Besides, if this
routine is used as object-method the filter is added immediately, so the
C<add_filter>-method must not be invoked separately.

=cut

sub create_filter {
	my $self = shift;
	my $filter;
	if ($self =~ /^Mensa::Dresden=HASH\(0x[\da-f]{7}\)/) {
		$filter = create_filter(@_);
		$self->add_filter($filter);
	} else {
		unshift @_, $self unless $self eq 'Mensa::Dresden';
		my ($criterion, $regex, $negative) = @_;
		$filter = Mensa::Dresden::Filter->new(
				$criterion, qr/$regex/i, $negative
		);
	}
	return $filter;
}

=item B<add_filter>

Adds the given filters to the canteen.

=cut

sub add_filter(@) {
	my $self = shift;
	for (@_) {
		carp("Not a filter: $_") and next
				unless /^Mensa::Dresden::Filter=HASH\(0x[\da-f]{7}\)/;
		my $filter_type = ($_->is_negative ? 'n' : 'p') . '_filters';
		push $self->{ $filter_type }, $_;
	}
}

=item B<reset_filters>

Removes all added filters.

=cut

sub reset_filters() {
	my $self = shift;
	$self->{p_filters} = [];
	$self->{n_filters} = [];
}

=back

=cut


1;

=head1 TODOs

=over 4

=item provide caching mechanism

If C<cache> is enabled the module will save the fetched html in a
temp-directory to enhance the speed of the module which otherwise
is dependent on the network-speed. The benefit will appear as the
offering is requested a second time.

=item extract the detail photo URL

=item check for network-connection

=back

=head1 SEE ALSO

For a more detailed usage of a filter see Dresden::Mensa::Filter.

http://www.studentenwerk-dresden.de/mensen/speiseplan/

=head1 AUTHOR

8ware, E<lt>8wared@googlemail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by 8ware

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.2 or,
at your option, any later version of Perl 5 you may have available.

=cut

__DATA__
<?xml version="1.0"?>
<xsl:stylesheet
		xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
		xmlns:fn="http://www.w3.org/2005/xpath-functions"
		xmlns:x="http://www.w3.org/1999/xhtml"
		version="1.0">
		<xsl:output method="xml" indent="yes" encoding="UTF-8" omit-xml-declaration="no"/>

	<xsl:param name="name"/>
	<xsl:variable name="base">http://www.studentenwerk-dresden.de/mensen/speiseplan/</xsl:variable>

	<xsl:template match="/">
		<xsl:element name="offering">
			<xsl:for-each select="//div[@id='spalterechtsnebenmenue']/table[@class='speiseplan'
					and thead/tr/th[@class='text']=$name]">
				<xsl:element name="mensa">
					<xsl:attribute name="name">
						<xsl:value-of select="thead/tr/th[@class='text']"/>
					</xsl:attribute>
					<xsl:for-each select="tbody/tr">
						<xsl:if test="string-length(td[@class='text']/a/@href) > 0">
							<xsl:element name="meal">
								<xsl:attribute name="url">
									<xsl:variable name="detailURI" select="td[@class='text']/a/@href"/>
									<!--xsl:value-of select="td[@class='text']/a/@href"/-->
									<xsl:choose>
										<xsl:when test="starts-with($detailURI, $base)">
											<xsl:value-of select="$detailURI"/>
										</xsl:when>
										<xsl:otherwise>
											<xsl:value-of select="concat($base, $detailURI)"/>
										</xsl:otherwise>
									</xsl:choose>
								</xsl:attribute>
								<xsl:element name="name">
									<xsl:value-of select="td[@class='text']/a"/>
								</xsl:element>
								<xsl:for-each select="td[@class='stoffe']/a/img">
									<xsl:element name="ingredient">
										<xsl:choose>
											<xsl:when test="contains(@title, 'vegan')">
												<xsl:value-of select="substring(@title, 10)"/>
											</xsl:when>
											<xsl:otherwise>
												<xsl:value-of select="substring(@title, 14)"/>
											</xsl:otherwise>
										</xsl:choose>
									</xsl:element>
								</xsl:for-each>
							</xsl:element>
						</xsl:if>
					</xsl:for-each>
				</xsl:element>
			</xsl:for-each>
		</xsl:element>
	</xsl:template>

</xsl:stylesheet>

